src/docs/sdk/performance/dynamic-sampling-context.mdx

```diff
@@ -0,0 +1,168 @@
+In the case that another vendor added Sentry values to an outgoing request, SDKs may overwrite those values.
+
+SDKs must not add other vendors' baggage from incoming requests to outgoing requests.
+Sentry SDKs only concern themselves with Sentry baggage.
+
+</Alert>
+
+The following is an example of what a baggage header containing Dynamic Sampling Context may look like:
+
+```
+baggage: other-vendor-value-1=foo;bar;baz, sentry-traceid=771a43a4192642f0b136d5159a501700, sentry-publickey=49d0f7386ad645858ae85020e393bef3; sentry-userid=Am%C3%A9lie, other-vendor-value-2=foo;bar;
+```
+
+See the [Payloads section](#payloads) for a complete list of key-value pairs that SDKs should propagate.
+
+## Payloads
+
+Dynamic Sampling Context is propagated via a baggage header and sent to Sentry via the `trace` envelope header.
+
+### Baggage-Header
+
+SDKs may set the following key-value pairs on baggage headers.
+While none of these values are strictly required, SDKs should make their best effort to add as many of them to the baggage header as possible when starting a trace.
+
+Some values are marked as "required for Dynamic sampling to work".
+Nothing bad is going to happen when these values are missing, however, Dynamic Sampling will not work without them.
+
+- `sentry-traceid` - **Required for Dynamic Sampling to work** - The original trace ID as generated by the SDK.
+- `sentry-publickey` - **Required for Dynamic Sampling to work** - Public key as defined by the user via the DSN in the SDK options.
+- `sentry-samplerate` - **Required for Dynamic Sampling to work** - Sample rate as defined by the user in the SDK options.
+- `sentry-release` - The release as defined by the user in the SDK options.
+- `sentry-environment` - The environment as defined by the user in the SDK options.
+- `sentry-transaction` - The name of the trace's origin transaction in unparameterized (raw) format.
+- `sentry-userid` - User ID as set by the user with <Link to="/sdk/unified-api/#scope">`scope.set_user`</Link>.
+- `sentry-usersegment` - User segment as set by the user with <Link to="/sdk/unified-api/#scope">`scope.set_user`</Link>.
+
+SDKs must set all of the keys in the form of "`sentry-[name]`".
+The prefix "`sentry-`" acts to identify key-value pairs set by Sentry SDKs.
+Additionally, we chose `[name]` to be written in "snake case" without any underscore ( `_` ) characters. This naming convention is the most language agnostic.
+
+### Envelope Header
+
+Dynamic Sampling Context is transferred to Sentry through the `trace` envelope header.
+The value of this header corresponds directly to the definition of <Link to="/sdk/performance/trace-context/#trace-context">Trace Context</Link>.
+
+When a transaction is reported to Sentry, the Dynamic Sampling Context must be mapped to the `trace` envelope header in the following way:
+
+- `sentry-traceid` ➝ `trace_id` - **required for Dynamic Sampling to work**
```

Comment:
why have this mapping? can't we have underscores in the headers? what if we had dashes in the headers and do a `.replace("-", "_")` when serializign to the envelope?

Reply:
Yup we thought about this too. We opted for this solution because we have to do some kind of mapping anyways because of the way the [user object](http://localhost:8000/sdk/performance/trace-context/#trace-context) is structured.

We cannot cleanly map from flat (header) structure to nested (envelope header). Ideally the user object wouldn't be an object on relay side but just two flat values, `user_id` and `user_segment`.

Does that kinda make sense?

Reply:
Ok, we just discussed this within the JS SDK team and we agree. We would also strongly prefer to be able to cleanly dump baggage into the envelope header - it's just way cleaner, allows for users to set baggage in the future, plus it allows us to save some bundle size in the SDK.

The problem we currently face is that relay accepts a nested user object and in baggage we have a flat structure:

```
// Let's assume we have have dashes in the baggage keys and do .replace("-", "_") before putting it into the envelope

// we want to dump this baggage header
baggage: sentry-user-id=adsf,sentry-user-segment=blah,sentry-public-key=foo

// into this trace envelope header
{
	public_key: "foo",
	user: {
		id: "adsf",
		segment: "blag"
	}
}
```

:D

![image](https://user-images.githubusercontent.com/8118419/175056497-2ebc74b9-8d28-4688-b1d5-e706607d782e.png)

It doesn't really work nicely as long `user` is an object. We propose lifting the `user` struct in Relay upwards into two separate values (`user_id` and `user_segment`).

TSC is coming up in 11 minutes. Let's discuss this there for a bit and then note the results here.

Reply:
As decided in the TSC:
- We will flatten the `user` object in relay into the fields `user_id` and `user_segment` (cc @jjbayer thanks for taking over this 🙇). We should also explore if it is possible to accept both, the nested `user` object and the flattened values - If yes the flattened values should have precedence.
- We will "dump" any values from incoming baggage headers onto the `trace` envelope header by stripping away `sentry-` (I will update the spec accordingly).

Let me know if I missed or misunderstood something.

Reply:
This relates to a comment I left on the doc elsewhere but `sentry-traceid` is redundant in the baggage since it's already also in `sentry-trace`. In that case we would want to however still pass it into the `trace` envelope header.